### ISSUE TITLE
PredictionFunction should not reuse the output object

### ISSUE THREAD

**Issue by CESARDELATORRE** [dotnet/machinelearning#1138]
`PredictionFunction` should create an instance of output object on every call.
The current behavior of reusing the object from call to call is confusing, and the performance gain is questionable.

(original bug description below)

### System information
- Using ML.NET v0.6 and PiGSTy/Static API

### Issue
We have migrated the sample using **TensorFlow as featurizer** and using the **pre-trained Inception v3 TF model**, then using an ML.NET SDCA learner for the classification.

This is the original LearningPipeline API-based sample that was working ok:
https://github.com/dotnet/machinelearning-samples/tree/master/samples/csharp/examples/DeepLearning_TensorFlowMLNETInceptionv3ModelScoring

The new code you can test is here:
repo: https://github.com/franperezlopez/machinelearning-samples
branch: features/samples-new-api

Here’s an screenshot showing the execution:

![image](https://user-images.githubusercontent.com/1712635/46432381-fccdae00-c702-11e8-84ed-794cde3d6398.png)

•	First, basically, the variables temp1 and temp2 and really the same object in memory (in Watcher, after "make object ID", you can see that both are pointing to the same position in memory. These variables come  from the object returned by **predictor.Predict()**. See the code in order to understand the code flow.

If `model.AsDynamic.MakePredictionFunction<ImageNetData, ImageNetStaticPrediction>(env).Predict(testData)` is meant to always return the same object in memory, it would be advisable to show that explicitly in the API, doing something like:

`model.AsDynamic.MakePredictionFunction<ImageNetData, ImageNetStaticPrediction>(env, testData).Predict()`

That way, testData is provided through the constructor instead as parameter for the function/method, so it is clearer that you can only use it “once”. But I'm not sure if the behavior is a bug, though...

Can you confirm if we need to “copy by value” the PredictionFunction to a new object whenever we want to do a different prediction with new test/sample data? For example, in a loop with multiple predictions, etc.

•	Second. The predictions we are getting are much worse in quality/probability than when using the LearningPipeline API. In the Watcher, looking at the metrics, you can see what’s going on. Basically, we’re exploring the preds.score returned by the SDCA, but it looks weird we’re getting numbers like **0.2** …?  --> When we were using the LearningPipeline for this same example we were getting probabilities around 0.999 or 0.0001.
o	You can also see in the screenshot the value for “**softmax2_pre_activation**”, which is the features vector coming from the neural network. 

![image](https://user-images.githubusercontent.com/1712635/46432660-a3b24a00-c703-11e8-85bd-93db619c348b.png)

Looks like the SDCA is not learning anything, it is always classifying as “broccoli”. So, we don’t know what’s going on here.
We might be doing something wrong or it might be a bug?

![image](https://user-images.githubusercontent.com/1712635/46432696-b7f64700-c703-11e8-8354-3c03a4448b80.png)

Bottom line, we’re blocked with this sample. It is not working properly with the new Static API after migrating a working sample with the LearningPipeline. 
Either we’re doing something wrong or there’s an issue/bug here?

Thanks,

**Comment by Zruty0** [comment]
Hi @CESARDELATORRE ,

First of all, you should not limit `maxIterations` to 4 if you want the model to be reasonable. This was done in the test in order to speed up training, but I'm pretty sure the model will be bad.

Yes, it looks like `PredictionFunction` reuses the OUTPUT object. Actually, it doesn't matter at all what is the input object, the output object returned will always be the same.

Historically, we had a parameter to `PredictionEngine` called `reuseRowObject` or something like that. If set to `true`, it will be reused from call to call, otherwise a new one will be created every time.

When writing `PredictionFunction`, I intended to drop this parameter, and make it always *not reuse* the output. It looks like either me or @TomFinley made a mistake and made the function always *reuse* the output. 

I think it's a confusing behavior that should be fixed.

cc: @TomFinley

**Comment by Zruty0** [comment]
To clarify, this bug is about fixing the reuse of the output object in `Predict` calls. I updated the description and name
